### DISCUSSION TITLE
NextJS 14: TypeError: req.on is not a function

### DISCUSSION THREAD

**Discussion by Nils19024** [vercel/next.js#64107] [Help] [answered]
### Summary

I have a API Route `/api/realtime/route.jsx` in NextJS 14:

```
import PocketBase from 'pocketbase';

const POCKET_BASE_URL = "http://192.168.230.28:8090";
const pb = new PocketBase(POCKET_BASE_URL);

export async function GET(req) {
    req.on('close', () => {
      console.log('Close event triggered');
    });

    const { readable, writable } = new TransformStream();
    const writer = writable.getWriter();

    pb.collection("user").subscribe(pb.authStore.model.id, change => {
      const message = `data: ${JSON.stringify([{type: "user"}, change])}\n\n`;
      writer.write(message);
    });

    const response = new Response(readable, {
      status: 200,
      headers: {
        "Content-Type": "text/event-stream",
      }
    });
    return response;
}
```

I want to close the server sent event connection to the client when `req.on('close', () => { });` is triggered. But i get the error message **TypeError: req.on is not a function**. Is this beacause of the new version NextJS 14? What do I need to change?

**Comment by icyJoseph** [comment]
Hi,

Yeah IIRC, the `req.on` was possible in Pages Router, because that extended the IncommingMessage API from Node.js, but in App Router, the request extends the WebAPI Request object. 

- https://developer.mozilla.org/en-US/docs/Web/API/Request

I am not sure how to accomplish what you want though. I'd guess that if the client closes the communication, then the stream gets dropped?

**Reply by Nils19024** [reply]
Yes that's what i want. I want to drop the the stream and the database event listener, when the client closes the connection, otherwise there will be a lot of database event listeners over time. Do you know how i can achieve that?

**Reply by icyJoseph** [reply]
Does the TransformStream on close event ever trigger?

**Reply by Nils19024** [reply, marked as answer]
I found the solution thanks to your link. It is:

```
request.signal.addEventListener("abort", () => {
      pb.collection('user').unsubscribe(); 
      pb.collection('orders_' + pb.authStore.model.club).unsubscribe();    
});    
```

So thanks a lot!

**Reply by icyJoseph** [reply]
Ah right! good call with that one! Glad to have helped.
